src/main/scala/chisel3/interface/Interface.scala

```diff
@@ -0,0 +1,121 @@
+  private[interface] def genModule(): Mod
+
+  /** Define how this module hooks up to the port-level interface. */
+  private[interface] def portMap: Seq[(Mod, Intf#Ports) => (Data, Data)]
+
+}
+
+/** Functionality which is common */
+sealed trait InterfaceCommon {
+
+  private[interface] type Ports <: Record
+
+  /** Returns the Record that is the port-level interface. */
+  private[interface] def ports(): Ports
+
+}
+
+/** A generator of different Interfaces. Currently, this is just an Interface
+  * that is not a Singleton.
+  */
+trait InterfaceGenerator extends InterfaceCommon
+
+/** An interface between hardware units. Any module that implements this
+  * interface may be separately compiled from any module that instantiates this
+  * interface.
+  */
+trait Interface extends InterfaceCommon { self: Singleton =>
+
+  /** This types represents the type of a valid conformance to this Interface.
+    */
+  private type Conformance[Mod <: RawModule] = ConformsTo[this.type, Mod]
+
+  /** The name of this interface. This will be used as the name of any module
+    * that implements this interface.
+    * I.e., this is the name of the `BlackBox` and `Module` that are provided
+    * below. The implementation of this method is just coming up with a good
+    * name derived from the class name. (The name of the Interface in FIRRTL
+    * will be the name of the Scala class that extends the Interface.)
+    */
+  private[interface] def interfaceName: String = {
+    val className = getClass().getName()
+    className
+      .drop(className.lastIndexOf('.') + 1)
+      .split('$')
+      .toSeq
+      .filter {
+        case str if str.forall(_.isDigit) => false
+        case str                          => true
+      }
+      .last
+  }
+
+  sealed trait Entity { this: BaseModule =>
+    val io: Ports
+  }
+
+  object Wrapper {
+
+    /** The black box that has the same ports as this interface. This is what is
+      * instantiated by any user of this interface, i.e., a test harness.
+      */
+    final class BlackBox extends chisel3.BlackBox with Entity {
+      final val io = IO(ports())
+
+      override final def desiredName = interfaceName
+    }
+
+    /** The module that wraps any module which conforms to this Interface.
+      */
+    final class Module[B <: RawModule](
+    )(
+      implicit conformance: Conformance[B])
+        extends RawModule
+        with Entity {
+      final val io = FlatIO(ports())
+
+      // Use a dummy clock and reset connection when constructing the module.
+      // This is fine as we rely on DataView to catch missing connections to
+      // clock and reset.  The dummy clock and reset will never be used.
+      private val internal = withClockAndReset(
+        WireInit(Clock(), DontCare),
+        WireInit(Reset(), DontCare)
+      ) {
+        chisel3.Module(conformance.genModule())
+      }
+
+      private implicit val pm = PartialDataView[B, Ports](
```

Comment:
Can we have some sort of less-strict "do your own wiring" alternative to at least support the Circt Chisel interface stub approach? Something like the `useStructuralConnect` option I prototyped but maybe more flexible?

Reply:
agree that this portMap style should be more a  convenience than a requirement...?

Reply:
I'm looking at a way that this can work. The specific problem (thanks for more side-band info, Daniel) is that we'd like to be able to connect an anonymous `Bundle` or multiple, nested anonymous `Bundle`s and make this pattern still usable. `DataView` does a check that classes match which makes this funky.

Daniel added a `useStructuralConnect` to this internally which does a weaker pairwise connection of the pairs defined in the `PortMap` that does not use `DataView`.

Reply:
This does work today, but it requires that the portMap is blown out into all the constituent connections. E.g., if you have an anonymous `Bundle` in the `Component`:

```scala
class Qux extends RawModule {
  val aa = IO(Input(new Bundle {
    val bb = Bool()
  }))
  val cc = IO(Output(new Bundle {
    val dd = Bool()
  }))

  cc.dd := aa.bb
}

class Foo extends RawModule {

  val a = IO(Input(Bool()))
  val b = IO(Output(Bool()))

  val qux1, qux2 = chisel3.Module(new BundleInterface.Wrapper.BlackBox)

  qux1.io.a.b := a
  qux2.io.a.b := qux1.io.c.d
  b := qux2.io.c.d
}
```

You need to blow out the connections in the `ConformsTo`:

```scala
implicit val quxConformance =
  new ConformsTo[BundleInterface.type, Qux] {

    override def genModule() = new Qux

    override def portMap = Seq(
      _.aa.bb -> _.a.b,
      _.cc.dd -> _.c.d
    )

  }
```

You cannot do the following:

```scala
```

Reply:
Instead of doing the blown-out connection, a second `DataView` could be used that defines the mapping between the structural, anonymous `Bundle` and the named `Bundle`. However, this is doing essentially the same work as blowing out the connections in the conformance.

@trilorez: Are you fine with keeping this more restrictive for now?

Reply:
Does this second `DataView` approach only handle a single level of bundles? In the "simple" example I'm working with, there are up to 4 levels of nested bundles.

Reply:
It will nest as far as you want. However, the information content is still going to be the same. The blown out connections are still there, just distributed one level at a time. Consider:

```scala
//> using scala "2.13.10"
//> using repository "https://s01.oss.sonatype.org/content/repositories/snapshots"
//> using lib "org.chipsalliance::chisel::5.0.0-M2"
//> using plugin "org.chipsalliance:::chisel-plugin::5.0.0-M2"
//> using options "-Ymacro-annotations"

import chisel3._
import chisel3.experimental.dataview._
import circt.stage.ChiselStage
import java.awt.MenuBar

class Baz extends Bundle {
  val a = Bool()
}

class Bar extends Bundle {
  val a = new Baz
}

class Foo extends Module {

  private val x = Wire(new Bar)
  private val y = Wire(new Bundle {
    val b = new Bundle {
      val b = Bool()
    }
  })

  implicit val view4 = PartialDataView[Bundle { val b: Bool }, Baz](
    _ => new Baz,
    _.b -> _.a
  )

  implicit val view3 =
    PartialDataView[Bundle { val b: Bundle { val b: Bool } }, Bar](
      _ => new Bar,
      _.b.viewAs[Baz] -> _.a
    )

  y.viewAs[Bar] :<>= x

}

object Main extends App {
  println(
    ChiselStage.emitCHIRRTL(
      new Foo,
      Array("--full-stacktrace")
    )
  )
}
```

Reply:
Ah, I see. Yeah, this works but it's cumbersome. It seems like more flexibility would be generally useful, at least to facilitate experimentation. But we can start here and add flexibility later.

Reply:
We can change this in the future either by relaxing how the connections work or by improvements to `DataView`. `DataView` already needs improvements around bit select and padding. We'll likely need even more power here. We can start with the strict approach and expand it as needed.